src/transactions.rs

```diff
@@ -0,0 +1,456 @@
+// Copyright (c) 2022 Espresso Systems (espressosys.com)
+// This file is part of the Seahorse library.
+
+// This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
+// This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
+// You should have received a copy of the GNU General Public License along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! The assets module.
+//!
+//! This module defines [Transaction], [TransactionEditor], and [Transactions], which provide CURD (create, read,
+//! update, and delete) operations, with the use of [KeyValueStore] to control the assets resource.
+
+use crate::{
+    key_value_store::*, AssetCode, KeystoreError, Ledger, RecordAmount, TransactionHash,
+    TransactionKind, TransactionReceipt, TransactionUID,
+};
+use chrono::{DateTime, Local};
+use jf_cap::{
+    keys::UserAddress,
+    structs::{ReceiverMemo, RecordOpening},
+    Signature,
+};
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+use std::collections::{BTreeMap, HashMap, HashSet};
+use std::ops::{Deref, DerefMut};
+
+/// A Transaction<L>with its UID as the primary key.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct Transaction<L: Ledger> {
+    /// Identifier for the Transaction, also the primary key in storage
+    uid: TransactionUID<L>,
+    /// Time when this transaction will expire if it's not completed
+    timeout: u64,
+    /// Hash which appears in the commited block for this transaction
+    hash: Option<TransactionHash<L>>,
+
+    // The uids of the outputs of this transaction for which memos have not yet been posted.
+    pending_uids: HashSet<u64>,
+
+    /// The accounts sending the transaction.
+    accounts: Vec<UserAddress>,
+    /// A receiver memo for each output, except for burned outputs.
+    memos: Vec<Option<ReceiverMemo>>,
+    sig: Signature,
+    /// If the transaction is a freeze, the expected frozen/unfrozen outputs.
+    freeze_outputs: Vec<RecordOpening>,
+    inputs: Vec<RecordOpening>,
+    outputs: Vec<RecordOpening>,
+
+    /// Time when this transaction was created in the transaction builder
```

Comment:
Technically it's the time the transaction was created or received. If you receive a transaction you don't know what time it was created.